### ISSUE TITLE
cURL cannot connect to my rustls-powerd https server

### ISSUE THREAD

**Issue by OverShifted** [rustls/rustls#1142]
I've created a simple https web server with the `tokio_rustls` and `tls_listener` crates. I am able to send requests to my server via Brave, Chromium, Firefox, wget, aria2c. But is seems like cURL and Luakit can't connect to my server.

Here is my code:
```rust
use std::sync::Arc;

use tokio::net::{TcpListener, TcpStream};
use tokio::io::{AsyncReadExt, AsyncWriteExt};

use tokio_rustls::rustls::ServerConfig;
use tokio_rustls::server::TlsStream;

use tls_listener::TlsListener;

// Based on https://github.com/tmccombs/tls-listener/blob/main/examples/tls_config/mod.rs
pub fn acceptor() -> tokio_rustls::TlsAcceptor {
    let key = {
        let file = std::fs::File::open("key.pem").unwrap();
        let mut reader = std::io::BufReader::new(file);
        let keys = rustls_pemfile::pkcs8_private_keys(&mut reader).unwrap();
        rustls::PrivateKey(keys[0].clone())
    };

    let cert = {
        let file = std::fs::File::open("cert.pem").unwrap();
        let mut reader = std::io::BufReader::new(file);
        rustls_pemfile::certs(&mut reader).unwrap().into_iter().map(rustls::Certificate).collect()
    };

    let mut config = ServerConfig::builder()
        .with_safe_defaults()
        .with_no_client_auth()
        .with_single_cert(cert, key)
        .unwrap();

    Arc::new(config).into()
}


#[tokio::main]
async fn main() {
    let listener = TcpListener::bind("0.0.0.0:8000").await.unwrap();
    let mut listener = TlsListener::new(acceptor(), listener);
    let mut processors = vec![];

    loop {
        let stream = listener.accept().await.unwrap();

        match stream {
            Err(error) => eprintln!("Error: {}", error),
            Ok(stream) => processors.push(tokio::spawn(async move {
                handle_connection(stream).await;
            }))
        };
    }
}

async fn handle_connection(mut socket: TlsStream<TcpStream>) {
    let mut buffer = [0; 1024];
    socket.read(&mut buffer).await.unwrap();
    let data = "<html><body><h1>Hello!</h1></body></html>";
    socket.write(format!("HTTP/1.1 200 OK\nContent-Length: {}\nContent-Type: text/html; charset=utf-8\n\n{}", data.len(), data).as_bytes()).await.unwrap();
    println!("Response sent.");
}
```

And here is how I've generated the `cert.pem` and `key.pem` files:
```
openssl genrsa -out key.pem
openssl req -new -key key.pem -out csr.pem
openssl x509 -req -days 9999 -in csr.pem -signkey key.pem -out cert.pem
rm csr.pem
```

wget's response:
```
$ wget --quiet --no-check-certificate https://127.0.0.1:8000 -O -
<html><body><h1>Hello!</h1></body></html>%
```

cURL's error:
```
$ curl -vk https://127.0.0.1:8000
*   Trying 127.0.0.1:8000...
* Connected to 127.0.0.1 (127.0.0.1) port 8000 (#0)
* ALPN: offers h2
* ALPN: offers http/1.1
* [CONN-0-0][CF-SSL] TLSv1.0 (OUT), TLS header, Certificate Status (22):
* [CONN-0-0][CF-SSL] TLSv1.3 (OUT), TLS handshake, Client hello (1):
* OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to 127.0.0.1:8000
* Closing connection 0
curl: (35) OpenSSL SSL_connect: SSL_ERROR_SYSCALL in connection to 127.0.0.1:8000
```

**Comment by ctz** [comment]
This example doesn't work because the private key does not match the certificate:
```
openssl genrsa -out key.pem
```
This command generates a PKCS#1 format RSA private key.
```
        let keys = rustls_pemfile::pkcs8_private_keys(&mut reader).unwrap();
```
This code reads a PKCS#8 format private key, and ignores other types present in the file.

This is why, when trying to reproduce this, the example panics with:
````
$ ./target/debug/testr 
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', src/main.rs:17:28
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Changing `rustls_pemfile::pkcs8_private_keys` to `rustls_pemfile::rsa_private_keys` in the above code fixes things for curl:

```
$ curl -k https://localhost:8000
<html><body><h1>Hello!</h1></body></html>

```

**Comment by OverShifted** [comment]
@ctz What is your openssl version? When I run `openssl genrsa -out key.pem`, I will end up with a PKCS#8 file (because it starts with `-----BEGIN PRIVATE KEY-----`, and not `-----BEGIN RSA PRIVATE KEY-----`).

In fact, when I use `rsa_private_keys`, I get the exact same panic message.

My openssl version is:
```
$ openssl version
OpenSSL 3.0.7 1 Nov 2022 (Library: OpenSSL 3.0.7 1 Nov 2022)
```

**Comment by OverShifted** [comment]
@ctz I tried using `openssl-1.1` (on Arch Linux) and `rustls_pemfile::rsa_private_keys` now works fine.
But it's still the same as before. Browsers, `openssl s_connect`, aria2c,.. work, but cURL doesn't.

**Comment by ctz** [comment]
I was testing on Ubuntu 20.04.5 which has:

```
$ openssl version
OpenSSL 1.1.1f  31 Mar 2020

$ curl --version
curl 7.68.0 (x86_64-pc-linux-gnu) libcurl/7.68.0 OpenSSL/1.1.1f zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3
Release-Date: 2020-01-08
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS brotli GSS-API HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL TLS-SRP UnixSockets
```

Could you take a packet capture of it going wrong?

**Comment by OverShifted** [comment]
Here you go: [captures.zip](https://github.com/rustls/rustls/files/10318508/captures.zip)
My "server" sends a FIN packet to curl for some reason.

**Comment by OverShifted** [comment]
Got it.
The problem is with Arch's curl build. Curl works fine when I run it inside a VM (Mint 20.3).
`curl --version`'s output on Mint is the exact same as yours.
`curl --version`'s output on Arch:
```
curl 7.87.0 (x86_64-pc-linux-gnu) libcurl/7.87.0 OpenSSL/3.0.7 zlib/1.2.13 brotli/1.0.9 zstd/1.5.2 libidn2/2.3.4 libpsl/0.21.1 (+libidn2/2.3.4) libssh2/1.10.0 nghttp2/1.51.0
Release-Date: 2022-12-21
Protocols: dict file ftp ftps gopher gophers http https imap imaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS brotli GSS-API HSTS HTTP2 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM NTLM_WB PSL SPNEGO SSL threadsafe TLS-SRP UnixSockets zstd
```
I will report this issue to Arch package maintainers after I find the related protocol/feature.

**Comment by ctz** [comment]
Ah, thanks! For the record replaying the ClientHello from curl.pcapng against rustls is successful, so nothing odd is going on at the TLS level.

**Comment by quininer** [comment]
I noticed that the problem doesn't occur if you don't use `tls-listener` crate, which I believe has some bugs in it.

like

```patch
     let listener = TcpListener::bind("0.0.0.0:8000").await.unwrap();
-    let mut listener = tls_listener::TlsListener::new(acceptor(), listener);
+    let acceptor = acceptor();
     let mut processors = vec![];

     loop {
-        let stream = acceptor.accept(stream).await;
+        let (stream, _) = listener.accept().await.unwrap();
         let stream = listener.accept().await.unwrap();
```

Also there are some problems in your code, for example you should use `flush` otherwise data may just be written to buffer and not sent, you should use `write_all` instead of `write`, otherwise only part of data may be written.

**Comment by quininer** [comment]
![Screenshot_20221229_211421](https://user-images.githubusercontent.com/6286761/209957716-95f30325-a042-4a33-af8a-09000ae3baec.png)

Continuing to investigate, it seems that this is not a bug, but rather a trigger for `tls-listener` default timeout. it seems that archlinux's curl handshake always takes more than 200ms.

**Comment by OverShifted** [comment]
@quininer Thank you for mentioning these problems. The scenario is much weirder than what I expected.
I copied the Mint (Ubuntu) curl binary to Arch, didn't work (maybe dynamic library problems?)
I compiled curl myself, didn't work.
I compiled curl myself with gnutls, didn't work.

Also, `SSL_ERROR_SYSCALL` isn't the only error that curl returns. Sometimes I get broken pipes (It seems to happen when I am loading another resource over the network, like reloading a tab or downloading a file which is really strange (possibly a kernel bug? because I am running Linux v6.0.12)) and rarely, connection reset by peers.

And, while I was rapidly running curls, I noticed that it actually worked once, but only once.
